Tests/Opc.Ua.Security.Certificates.Tests/CertificateTests.cs

```diff
@@ -0,0 +1,381 @@
+using System.Text;
+using NUnit.Framework;
+
+namespace Opc.Ua.Security.Certificates.Tests
+{
+    /// <summary>
+    /// Tests for the CertificateFactory class.
+    /// </summary>
+    [TestFixture, Category("Certificate")]
+    [Parallelizable]
+    [SetCulture("en-us")]
+    public class CertificateTests
+    {
+        #region DataPointSources
+        public const string Subject = "CN=Test Cert Subject";
+
+        [DatapointSource]
+        public CertificateAsset[] CertificateTestCases = new AssetCollection<CertificateAsset>(TestUtils.EnumerateTestAssets("*.?er")).ToArray();
+
+        [DatapointSource]
+        public KeyHashPair[] KeyHashPairs = new KeyHashPairCollection {
+            { 1024, HashAlgorithmName.SHA256 /* SHA-1 is deprecated HashAlgorithmName.SHA1*/ },
+            { 2048, HashAlgorithmName.SHA256 },
+            { 3072, HashAlgorithmName.SHA384 },
+            { 4096, HashAlgorithmName.SHA512 } }.ToArray();
+
+#if !NET462
+        [DatapointSource]
+        public ECCurve[] NamedCurves = typeof(ECCurve.NamedCurves).GetProperties(BindingFlags.Public | BindingFlags.Static).Select(x => ECCurve.CreateFromFriendlyName(x.Name)).ToArray();
+#endif
+        #endregion
+
+        #region Test Setup
+        /// <summary>
+        /// Set up a Global Discovery Server and Client instance and connect the session
+        /// </summary>
+        [OneTimeSetUp]
+        protected void OneTimeSetUp()
+        {
+        }
+
+        /// <summary>
+        /// Clean up the Test PKI folder
+        /// </summary>
+        [OneTimeTearDown]
+        protected void OneTimeTearDown()
+        {
+        }
+        #endregion
+
+        #region Test Methods
+        /// <summary>
+        /// Verify self signed app certs.
+        /// </summary>
+        [Test]
+        public void VerifyOneSelfSignedAppCertForAll()
+        {
+            var builder = CertificateBuilder.Create(Subject)
+                .SetNotBefore(DateTime.Today.AddYears(-1))
+                .SetNotAfter(DateTime.Today.AddYears(25))
+                .AddExtension(new X509SubjectAltNameExtension("urn:opcfoundation.org:mypc", new string[] { "mypc", "mypc.opcfoundation.org", "192.168.1.100" }));
+            foreach (var keyHash in KeyHashPairs)
+            {
+                var cert = builder
+                    .SetHashAlgorithm(keyHash.HashAlgorithmName)
+                    .SetRSAKeySize(keyHash.KeySize)
+                    .CreateForRSA();
+                Assert.NotNull(cert);
+                WriteCertificate(cert, $"Default cert with RSA {keyHash.KeySize} {keyHash.HashAlgorithmName} signature.");
+                Assert.AreEqual(keyHash.HashAlgorithmName, Oids.GetHashAlgorithmName(cert.SignatureAlgorithm.Value));
+            }
+        }
+
+        [Theory]
+        public void CreateSelfSignedForRSATests(
+            KeyHashPair keyHashPair
+            )
+        {
+            // default cert with custom key
```

Comment:
You could create separate test cases (with names build from code comments), so if something is failing it is easier to detect on the name, were was the issue

Reply:
done